### PR TITLE
Add types to parameter.py.

### PR DESCRIPTION
**Pull request by InvincibleRMC**

### CODE REVIEW

File: rclpy/rclpy/parameter.py [hunk 1]
```diff
@@ -28,18 +29,19 @@
...
-                                    List[float], Tuple[float, ...], array.array[float],
-                                    List[str], Tuple[str, ...], array.array[str]]
+AllowableParameterValue = Union[None, bool, int, float, str,
+                                List[bytes], Tuple[bytes, ...],
+                                List[bool], Tuple[bool, ...],
+                                List[int], Tuple[int, ...], 'array.array[int]',
+                                List[float], Tuple[float, ...], 'array.array[float]',
+                                List[str], Tuple[str, ...], 'array.array[str]']
 
+AllowableParameterValueT = TypeVar('AllowableParameterValueT', AllowableParameterValue, None)
```
[the remark this answers is not in the document]
Comment (InvincibleRMC) [reply rclpy/rclpy/parameter.py:39]: The reason this was done to suppress a mypy error which incorrectly complains about the `AllowableParameterValueT` being given a default value. Pyright for example correctly infers the type and does not complain. Another solution would be `Optional[AllowableParameterValueT]`  in the `Parameter` constructor.

[the remark this answers is not in the document]
Comment (InvincibleRMC) [reply rclpy/rclpy/parameter.py]: I did a little cleanup around this and linked to the open mypy [issue](https://github.com/python/mypy/issues/3737).

**Review by sloretz — APPROVED** [review]
Thank you for the PR!

**Comment by sloretz** [comment]
Gist: https://gist.githubusercontent.com/sloretz/914d83d9692ee5a19b612230bfbb0a1f/raw/ad37a28a261a83b1d57b1191b06fd7aae064cba4/ros2.repos
BUILD args:  --packages-up-to rclpy
TEST args:  --packages-select rclpy
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/14355
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/21607/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/15895/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/1112/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22325/)

**Pull request merged**